gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/CommitStepWorkflowImpl.java

```diff
@@ -59,10 +61,19 @@ public class CommitStepWorkflowImpl implements CommitStepWorkflow {
   @Override
   public CommitStats commit(WUProcessingSpec workSpec) {
     CommitStats commitGobblinStats = activityStub.commit(workSpec);
-    TemporalEventTimer.Factory timerFactory = new TemporalEventTimer.Factory(workSpec.getEventSubmitterContext());
-    timerFactory.create(TimingEvent.LauncherTimings.JOB_SUMMARY)
-        .withMetadata(TimingEvent.DATASET_TASK_SUMMARIES, GsonUtils.GSON_WITH_DATE_HANDLING.toJson(convertDatasetStatsToTaskSummaries(commitGobblinStats.getDatasetStats())))
-        .submit();
+
+    if(!commitGobblinStats.getOptFailure().isPresent() || commitGobblinStats.getNumCommittedWorkUnits() > 0) {
+      TemporalEventTimer.Factory timerFactory = new TemporalEventTimer.Factory(workSpec.getEventSubmitterContext());
+      timerFactory.create(TimingEvent.LauncherTimings.JOB_SUMMARY)
+          .withMetadata(TimingEvent.DATASET_TASK_SUMMARIES, GsonUtils.GSON_WITH_DATE_HANDLING.toJson(
+              convertDatasetStatsToTaskSummaries(commitGobblinStats.getDatasetStats())))
+          .submit();
+    }
+    if(commitGobblinStats.getOptFailure().isPresent()){
+      throw ApplicationFailure.newNonRetryableFailureWithCause(
+          String.format("Failed to commit dataset state for some dataset(s)"), FailedDatasetUrnsException.class.toString(),
+          commitGobblinStats.getOptFailure().get());
+    }
```

Comment:
this may not be type-safe at runtime, if the type of `getOptFailure` is not `FailedDatasetUrnsException`.  if that's the only possibility, then change the field type in `CommitStats`.  otherwise:
```
  Exception failure = commitGobblinStats.getOptFailure().get();
  throw ApplicationFailure.newNonRetryableFailureWithCause(
      ...,
      failure.getClass().getName(),
      failure);
```